src/PersistorHelper.js

```diff
@@ -0,0 +1,186 @@
+const Crypto = require('crypto')
+const Stream = require('stream')
+const Settings = require('./Settings')
+const Logger = require('logger-sharelatex')
+const { WriteError, ReadError, NotFoundError } = require('./Errors')
+const { promisify } = require('util')
+
+const pipeline = promisify(Stream.pipeline)
+
+// Observes data that passes through and computes some metadata for it
+// - specifically, it computes the number of bytes transferred, and optionally
+//   computes a cryptographic hash based on the 'hash' option. e.g., pass
+//   { hash: 'md5' } to compute the md5 hash of the stream
+// - if 'metric' is supplied as an option, this metric will be incremented by
+//   the number of bytes transferred
+class ObserverStream extends Stream.Transform {
+  constructor(options) {
+    options.autoDestroy = true
+    super(options)
+
+    this.bytes = 0
+
+    if (options.hash) {
+      this.hash = Crypto.createHash(options.hash)
+    }
+
+    if (options.metric && Settings.Metrics) {
+      const onEnd = () => {
+        Settings.Metrics.count(options.metric, this.bytes)
+      }
+      this.once('error', onEnd)
+      this.once('end', onEnd)
+    }
+  }
+
+  _transform(chunk, encoding, done) {
+    if (this.hash) {
+      this.hash.update(chunk)
+    }
+    this.bytes += chunk.length
+    this.push(chunk)
+    done()
+  }
+
+  getHash() {
+    return this.hash && this.hash.digest('hex')
+  }
+}
+
+module.exports = {
+  ObserverStream,
+  calculateStreamMd5,
+  verifyMd5,
+  getReadyPipeline,
+  wrapError,
+  hexToBase64,
+  base64ToHex
+}
+
+// returns a promise which resolves with the md5 hash of the stream
+// - consumes the stream
+function calculateStreamMd5(stream) {
+  const hash = Crypto.createHash('md5')
+  hash.setEncoding('hex')
+
+  return pipeline(stream, hash).then(() => hash.read())
+}
```

Comment:
async/await version:
```suggestion
async function calculateStreamMd5(stream) {
  const hash = Crypto.createHash('md5')
  hash.setEncoding('hex')

  await pipeline(stream, hash)
  return hash.read()
}
```